File: TripKitUI/views/TKUIStyledButton.m
//
//  TKUIStyledButton.m
//  TripKit
//
//  Created by Kuan Lun Huang on 30/03/2015.
//
//

#import "TKUIStyledButton.h"

#ifdef TK_NO_MODULE
#import "TripKit.h"
#import "TripKit/TripKit-Swift.h"
#else
@import TripKit;
#import "TripKitUI/TripKitUI-Swift.h"
#endif

#import "TKStyleManager+TripKitUI.h"

@implementation TKUIStyledButton

- (instancetype)init
{
  self = [super init];
  
  if (self) {
    [self commonInit];
  }
  
  return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
  self = [super initWithCoder:aDecoder];
  
  if (self) {
    [self commonInit];
  }
  
  return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
  self = [super initWithFrame:frame];
  
  if (self) {
    [self commonInit];
  }
  
  return self;
}

#pragma mark - Private methods

- (void)commonInit
{
  NSString *style = [[self.titleLabel.font fontDescriptor] fontAttributes][UIFontDescriptorTextStyleAttribute];
  
  if (! style) {
    CGFloat point = self.titleLabel.font.pointSize;
    self.titleLabel.font = [TKStyleManager systemFontWithSize:point];
  } else {
    self.titleLabel.font = [TKStyleManager customFontForTextStyle:style];
  }
}

@end

File: TripKitUI/views/TKUIStyledButton.swift
//
//  TKUIStyledButton.swift
//  TripKitUI-iOS
//
//  Created by Adrian Schönig on 31/1/21.
//  Copyright © 2021 SkedGo Pty Ltd. All rights reserved.
//

import UIKit
import TripKit

@objc
open class TKUIStyledButton: UIButton {
  public override init(frame: CGRect) {
    super.init(frame: frame)
    didInit()
  }
  
  public required init?(coder: NSCoder) {
    super.init(coder: coder)
    didInit()
  }
  
  private func didInit() {
    guard let label = titleLabel else { return }
    
    if let style = label.font.fontDescriptor.fontAttributes[.textStyle] as? String {
      label.font = TKStyleManager.customFont(forTextStyle: UIFont.TextStyle(rawValue: style))
    } else {
      label.font = TKStyleManager.systemFont(size: label.font.pointSize)
    }
  }
}
